File: CBORTest/BEncodingTest.cs
using System;
using System.IO;
using NUnit.Framework;
using PeterO;
using PeterO.Cbor;

namespace Test {
  [TestFixture]
  public class BEncodingTest {
    private static CBORObject EncodingFromBytes(byte[] b) {
      try {
        using var s = new Test.DelayingStream(b);
        return BEncoding.Read(s);
      } catch (IOException ex) {
        throw new CBORException(String.Empty, ex);
      }
    }

    private static byte[] EncodingToBytes(CBORObject b) {
      try {
        using var ms = new Test.DelayingStream();
        BEncoding.Write(b, ms);
        return ms.ToArray();
      } catch (IOException ex) {
        throw new CBORException(String.Empty, ex);
      }
    }

    public static void DoTestLong(long value) {
      string b = "i" + TestCommon.LongToString(value) + "e";
      CBORObject beo = EncodingFromBytes(DataUtilities.GetUtf8Bytes(b,
            false));
      Assert.AreEqual(value, beo.AsNumber().ToInt64Checked());
      string newb = DataUtilities.GetUtf8String(EncodingToBytes(beo), false);
      Assert.AreEqual(b, newb);
    }

    public static void DoTestString(string value) {
      string b = DataUtilities.GetUtf8Length(value, false) + ":" + value;
      CBORObject beo = EncodingFromBytes(DataUtilities.GetUtf8Bytes(b,
            false));
      Assert.AreEqual(value, beo.AsString());
      string newb = DataUtilities.GetUtf8String(EncodingToBytes(beo), false);
      Assert.AreEqual(b, newb);
    }

    [Test]
    public void TestLong() {
      DoTestLong(0);
      DoTestLong(-1);
      DoTestLong(Int32.MinValue);
      DoTestLong(Int32.MaxValue);
      DoTestLong(Int64.MinValue);
      DoTestLong(Int64.MaxValue);
    }

    [Test]
    public void TestList() {
      var beo = CBORObject.NewArray();
      _ = beo.Add(ToObjectTest.TestToFromObjectRoundTrip(1));
      _ = beo.Add(ToObjectTest.TestToFromObjectRoundTrip("two"));
      _ = beo.Add(ToObjectTest.TestToFromObjectRoundTrip(3));
      _ = beo.Add(ToObjectTest.TestToFromObjectRoundTrip("four"));
      Assert.AreEqual(4, beo.Count);
      Assert.AreEqual(1, beo[0].AsNumber().ToInt64Checked());
      {
        string stringTemp = beo[1].AsString();
        Assert.AreEqual(
          "two",
          stringTemp);
      }
      Assert.AreEqual(3, beo[2].AsNumber().ToInt64Checked());
      {
        string stringTemp = beo[3].AsString();
        Assert.AreEqual(
          "four",
          stringTemp);
      }
      byte[] b = EncodingToBytes(beo);
      beo = EncodingFromBytes(b);
      Assert.AreEqual(4, beo.Count);
      Assert.AreEqual(1, beo[0].AsNumber().ToInt64Checked());
      {
        string stringTemp = beo[1].AsString();
        Assert.AreEqual(
          "two",
          stringTemp);
      }
      Assert.AreEqual(3, beo[2].AsNumber().ToInt64Checked());
      {
        string stringTemp = beo[3].AsString();
        Assert.AreEqual(
          "four",
          stringTemp);
      }
    }

    [Test]
    public void TestDictionary() {
      var beo = CBORObject.NewMap();
      beo["zero"] = ToObjectTest.TestToFromObjectRoundTrip(1);
      beo["one"] = ToObjectTest.TestToFromObjectRoundTrip("two");
      beo["two"] = ToObjectTest.TestToFromObjectRoundTrip(3);
      beo["three"] = ToObjectTest.TestToFromObjectRoundTrip("four");
      Assert.AreEqual(4, beo.Count);
      Assert.AreEqual(1, beo["zero"].AsNumber().ToInt64Checked());
      {
        string stringTemp = beo["one"].AsString();
        Assert.AreEqual(
          "two",
          stringTemp);
      }
      Assert.AreEqual(3, beo["two"].AsNumber().ToInt64Checked());
      {
        string stringTemp = beo["three"].AsString();
        Assert.AreEqual(
          "four",
          stringTemp);
      }
      byte[] b = EncodingToBytes(beo);
      beo = EncodingFromBytes(b);
      Assert.AreEqual(4, beo.Count);
      Assert.AreEqual(1, beo["zero"].AsNumber().ToInt64Checked());
      {
        string stringTemp = beo["one"].AsString();
        Assert.AreEqual(
          "two",
          stringTemp);
      }
      Assert.AreEqual(3, beo["two"].AsNumber().ToInt64Checked());
      {
        string stringTemp = beo["three"].AsString();
        Assert.AreEqual(
          "four",
          stringTemp);
      }
    }

    [Test]
    public void TestString() {
      DoTestString(String.Empty);
      DoTestString(" ");
      DoTestString("test");

      DoTestString(
        TestCommon.Repeat("three", 15));
      DoTestString("te\u007fst");
      DoTestString("te\u0080st");
      DoTestString("te\u3000st");
      DoTestString("te\u07ffst");
      DoTestString("te\u0800st");
      DoTestString("te\uffffst");
      DoTestString("te\ud7ffst");
      DoTestString("te\ue000st");
      DoTestString("te\ud800\udc00st");
      DoTestString("te\udbff\udc00st");
      DoTestString("te\ud800\udfffst");
      DoTestString("te\udbff\udfffst");
    }
  }
}

File: src/test/java/com/upokecenter/test/BEncodingTest.java
package com.upokecenter.test;

import java.io.*;

import org.junit.Assert;
import org.junit.Test;
import com.upokecenter.util.*;
import com.upokecenter.cbor.*;

  public class BEncodingTest {
    private static CBORObject EncodingFromBytes(byte[] b) {
      try {
        java.io.ByteArrayInputStream s = null;
try {
s = new java.io.ByteArrayInputStream(b);

          return BEncoding.Read(s);
}
finally {
try { if (s != null)s.close(); } catch (java.io.IOException ex) {}
}
      } catch (IOException ex) {
        throw new CBORException("", ex);
      }
    }

    private static byte[] EncodingToBytes(CBORObject b) {
      try {
        java.io.ByteArrayOutputStream s = null;
try {
s = new java.io.ByteArrayOutputStream();

          BEncoding.Write(b, s);
          return s.toByteArray();
}
finally {
try { if (s != null)s.close(); } catch (java.io.IOException ex) {}
}
      } catch (IOException ex) {
        throw new CBORException("", ex);
      }
    }

    public static void doTestLong(long value) {
      String b = "i" + value + "e";
      CBORObject beo = EncodingFromBytes(com.upokecenter.util.DataUtilities.GetUtf8Bytes(b, true));
      Assert.assertEquals(value, beo.AsInt64());
      String newb = com.upokecenter.util.DataUtilities.GetUtf8String(EncodingToBytes(beo), true);
      Assert.assertEquals(b, newb);
    }

    public static void doTestString(String value) {
      String b = DataUtilities.GetUtf8Length(value, false) + ":" + value;
      CBORObject beo = EncodingFromBytes(com.upokecenter.util.DataUtilities.GetUtf8Bytes(b, true));
      Assert.assertEquals(value, beo.AsString());
      String newb = com.upokecenter.util.DataUtilities.GetUtf8String(EncodingToBytes(beo), true);
      Assert.assertEquals(b, newb);
    }

    @Test
    public void testLong() {
      doTestLong(0);
      doTestLong(-1);
      doTestLong(Integer.MIN_VALUE);
      doTestLong(Integer.MAX_VALUE);
      doTestLong(Long.MIN_VALUE);
      doTestLong(Long.MAX_VALUE);
    }

    @Test
    public void testList() {
      CBORObject beo = CBORObject.NewArray();
      beo.Add(CBORObject.FromObject(1));
      beo.Add(CBORObject.FromObject("two"));
      beo.Add(CBORObject.FromObject(3));
      beo.Add(CBORObject.FromObject("four"));
      Assert.assertEquals(4, beo.size());
      Assert.assertEquals(1, beo.get(0).AsInt64());
      Assert.assertEquals("two", beo.get(1).AsString());
      Assert.assertEquals(3, beo.get(2).AsInt64());
      Assert.assertEquals("four", beo.get(3).AsString());
      byte[] b = EncodingToBytes(beo);
      beo = EncodingFromBytes(b);
      Assert.assertEquals(4, beo.size());
      Assert.assertEquals(1, beo.get(0).AsInt64());
      Assert.assertEquals("two", beo.get(1).AsString());
      Assert.assertEquals(3, beo.get(2).AsInt64());
      Assert.assertEquals("four", beo.get(3).AsString());
    }

    @Test
    public void testDictionary() {
      CBORObject beo = CBORObject.NewMap();
      beo.set("zero",CBORObject.FromObject(1));
      beo.set("one",CBORObject.FromObject("two"));
      beo.set("two",CBORObject.FromObject(3));
      beo.set("three",CBORObject.FromObject("four"));
      Assert.assertEquals(4, beo.size());
      Assert.assertEquals(1, beo.get("zero").AsInt64());
      Assert.assertEquals("two", beo.get("one").AsString());
      Assert.assertEquals(3, beo.get("two").AsInt64());
      Assert.assertEquals("four", beo.get("three").AsString());
      byte[] b = EncodingToBytes(beo);
      beo = EncodingFromBytes(b);
      Assert.assertEquals(4, beo.size());
      Assert.assertEquals(1, beo.get("zero").AsInt64());
      Assert.assertEquals("two", beo.get("one").AsString());
      Assert.assertEquals(3, beo.get("two").AsInt64());
      Assert.assertEquals("four", beo.get("three").AsString());
    }

    @Test
    public void testString() {
      doTestString("");
      doTestString(" ");
      doTestString("test");

  doTestString("testoifdoifdodfioidfifdidfoiidofiosidoiofdsoiiofdsiofdiosiodfiosdoiffiodsiosdfiods"
);
      doTestString("te\u007fst");
      doTestString("te\u0080st");
      doTestString("te\u3000st");
      doTestString("te\u07ffst");
      doTestString("te\u0800st");
      doTestString("te\uffffst");
      doTestString("te\ud7ffst");
      doTestString("te\ue000st");
      doTestString("te\ud800\udc00st");
      doTestString("te\udbff\udc00st");
      doTestString("te\ud800\udfffst");
      doTestString("te\udbff\udfffst");
    }
  }
